components/org.wso2.carbon.identity.recovery/src/main/java/org/wso2/carbon/identity/recovery/signup/UserSelfRegistrationManager.java

```diff
@@ -1476,4 +1488,159 @@ private void auditRecoveryConfirm(UserRecoveryData recoveryData, String errorMsg
+        }
+        if (StringUtils.isNotBlank(verifiedChannelType)) {
+            properties.put(IdentityRecoveryConstants.SELF_REGISTRATION_VERIFIED_CHANNEL, verifiedChannelType);
+        }
+        if (StringUtils.isNotBlank(verifiedChannelClaim)) {
+            properties.put(IdentityRecoveryConstants.SELF_REGISTRATION_VERIFIED_CHANNEL_CLAIM, verifiedChannelClaim);
+        }
+        if (metaProperties != null) {
+            for (Map.Entry<String, String> entry : metaProperties.entrySet()) {
+                if (StringUtils.isNotBlank(entry.getValue()) && StringUtils.isNotBlank(entry.getKey())) {
+                    properties.put(entry.getKey(), entry.getValue());
+                }
+            }
+        }
+
+        Event identityMgtEvent = new Event(eventName, properties);
+        try {
+            IdentityRecoveryServiceDataHolder.getInstance().getIdentityEventService().handleEvent(identityMgtEvent);
+        } catch (IdentityEventException e) {
+            log.error("Error occurred while publishing event " + eventName + " for user " + user);
+            throw Utils.handleServerException(IdentityRecoveryConstants.ErrorMessages.ERROR_CODE_PUBLISH_EVENT,
+                    eventName, e);
+        }
+
+    }
+
+    /**
+     * Method to publish pre self sign up confirm event.
+     *
+     * @param code self signup confirmation code
+     * @param verifiedChannelType verified channel type
+     * @param verifiedChannelClaim verified channel claim.
+     * @param metaProperties metaproperties of the request
+     * @param eventName event name (PRE_SELF_SIGNUP_CONFIRM)
+     * @throws IdentityRecoveryException
+     */
+    private void publishEvent(String code, String verifiedChannelType,String verifiedChannelClaim,
+                              Map<String, String> metaProperties,
+                              String eventName) throws
+            IdentityRecoveryException {
+
+        HashMap<String, Object> properties = new HashMap<>();
+
+        if (StringUtils.isNotBlank(code)) {
+            properties.put(IdentityRecoveryConstants.SELF_REGISTRATION_CODE, code);
+        }
+        if (StringUtils.isNotBlank(verifiedChannelType)) {
+            properties.put(IdentityRecoveryConstants.SELF_REGISTRATION_VERIFIED_CHANNEL, verifiedChannelType);
+        }
+        if (StringUtils.isNotBlank(verifiedChannelClaim)) {
+            properties.put(IdentityRecoveryConstants.SELF_REGISTRATION_VERIFIED_CHANNEL_CLAIM, verifiedChannelClaim);
+        }
+        if (metaProperties != null) {
+            for (Map.Entry<String, String> entry : metaProperties.entrySet()) {
+                if (StringUtils.isNotBlank(entry.getValue()) && StringUtils.isNotBlank(entry.getKey())) {
+                    properties.put(entry.getKey(), entry.getValue());
+                }
+            }
+        }
+
+        Event identityMgtEvent = new Event(eventName, properties);
```

Comment:
can extract to a private method to avoid duplications

Reply:
done with 28f7703e01d2c0eceb0a2d65929740628425880a